packages/datastore/src/datastore.ts

```diff
@@ -343,48 +362,60 @@ class Datastore implements IIterable<Table<Schema>>, IMessageHandler, IDisposabl
   private constructor(
     context: Datastore.Context,
     tables: BPlusTree<Table<Schema>>,
-    broadcastHandler?: IMessageHandler,
+    adapter?: IServerAdapter,
     transactionIdFactory?: Datastore.TransactionIdFactory
   ) {
     this._context = context;
     this._tables = tables;
-    this._broadcastHandler = broadcastHandler || null;
+    this._adapter = adapter || null;
     this._transactionIdFactory = transactionIdFactory || createDuplexId;
+    if (this._adapter) {
+      this._adapter.received.connect(
+        this._onRemoteTransaction,
+        this
+      );
+    }
   }
 
+  /**
+   * Handle a transaction from the server adapter.
+   */
+  private _onRemoteTransaction(
+    sender: IServerAdapter,
+    args: IServerAdapter.IReceivedArgs
+  ): void {
+    this._processTransaction(args);
+  }
 
   /**
    * Apply a transaction to the datastore.
    *
-   * @param transaction - The data of the transaction.
+   * @param transactionApplication - The data of the transaction.
    *
-   * @throws An exception if `apply` is called during a mutation.
+   * @throws An exception if `processTransaction` is called during a mutation.
    *
    * #### Notes
    * If changes are made, the `changed` signal will be emitted.
    */
-  private _applyTransaction(transaction: Datastore.Transaction, fromQueue=false): void {
-    if (!this._transactionQueue.isEmpty && !fromQueue) {
```

Comment:
@vidartf I've restored most of the logic regarding queuing patches if they arrive in the middle of an existing transaction. However, I've been less strict about requiring that they are queued in order, since the cemetery should take care of that. In particular, I have removed this logic about automatically queuing if there is already something in the queue.

Does that seem sensible to you?

Reply:
Seems good!